### PR TITLE
Replace BigNumber and Decimal with SafeDecimal

### PR DESCRIPTION
**Pull request by tiagofilipenunes**
- [x] Create SafeDecimal class
- [x] Add tests to compare functionality with BigNumber
- [x] Replace BigNumber with SafeDecimal
- [x] Replace Decimal with BigNumber (apart from in the sdk)
- [x] Solve absolute paths issue when running vitest (can't find import when running yarn test and using absolute paths)
- [x] Remove passing '...' to SafeDecimal
  - As both the TokenInputField which is used throughout the app and the carbonSDK.getTradeData take a value/sourceInput/targetInput as text, the simplest solution would be to set them to 'NaN', which in my view is as much value as leaving it to '...'. With SafeDecimal, there is now no unexpected behavior when dealing with strings that can't be parsed to number.

Others
- Minor fixes to tradeWidgetBuySell
  - No Liquidity still showed the header in the widget for the buy section (shouldn't show)
  - Fix GTM tags to use targetAmount when there is a targetError
- Orderbook refactor
  - Unnecessary checks removed

### CODE REVIEW

**Review by zavelevsky — CHANGES_REQUESTED** [review]

File: src/libs/safedecimal/index.ts [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+    }
+  } else {
+    return value;
+  }
+}
+
+class SafeDecimal extends Decimal {
+  constructor(value: Decimal.Value) {
+    super(sanitizeDecimal(value)); // will do super('NaN') if value is not parseable
+    this.constructor = SafeDecimal;
```
[the remark this answers is not in the document]
Comment (tiagofilipenunes) [reply src/libs/safedecimal/index.ts:20]: Unfortunately do, this is a problem with how decimal.js is written, as it overrides the constructor on every instance to be Decimal [line 4,290 in decimal.mjs (x.constructor = Decimal)].

I've added a comment to make it obvious

File: src/libs/safedecimal/index.ts [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+      return trimmedNumber;
+    } else {
+      return 'NaN';
+    }
+  } else {
+    return value;
+  }
+}
+
+class SafeDecimal extends Decimal {
```
[the remark this answers is not in the document]
Comment (tiagofilipenunes) [reply src/libs/safedecimal/index.ts:17]: Did this, and realised that SafeDecimal doesn't cover Octal, Binary, Hexadecimal and special cases (Infinity and NaN directly). Changes made resulted in the following output with `npm test`

```
 Testing decimal.js

 Testing absoluteValue... 100 of 100 tests passed in 1 ms
 Testing acos... 50 of 50 tests passed in 77 ms
 Testing acosh... 50 of 50 tests passed in 14 ms
 Testing asin... 50 of 50 tests passed in 48 ms
 Testing asinh... 50 of 50 tests passed in 14 ms
 Testing atan... 50 of 50 tests passed in 58 ms
 Testing atan2... 100 of 100 tests passed in 31 ms
 Testing atanh... 50 of 50 tests passed in 19 ms
 Testing cubeRoot... 50 of 50 tests passed in 27 ms
 Testing ceil... 50 of 50 tests passed in 0 ms
 Testing clamp... 29 of 29 tests passed in 0 ms
 Testing clone... 65 of 65 tests passed in 1 ms
 Testing comparedTo... 1000 of 1000 tests passed in 3 ms
 Testing config... 206 of 206 tests passed in 1 ms
 Testing cos... 101 of 101 tests passed in 24 ms
 Testing cosh... 100 of 100 tests passed in 33 ms
 Testing Decimal...
  Test number 860 failed: assertException
  Expected: ' NaN' to raise a DecimalError.
  Actual:   no exception, result: NaN
  Test number 861 failed: assertException
  Expected: 'NaN ' to raise a DecimalError.
  Actual:   no exception, result: NaN
  Test number 862 failed: assertException
  Expected: ' NaN ' to raise a DecimalError.
  Actual:   no exception, result: NaN
  Test number 863 failed: assertException
  Expected: ' -NaN' to raise a DecimalError.
  Actual:   no exception, result: NaN
  Test number 864 failed: assertException
  Expected: ' +NaN' to raise a DecimalError.
  Actual:   no exception, result: NaN
  Test number 865 failed: assertException
  Expected: '-NaN ' to raise a DecimalError.
  Actual:   no exception, result: NaN
  Test number 866 failed: assertException
  Expected: '+NaN ' to raise a DecimalError.
  Actual:   no exception, result: NaN
  Test number 874 failed: assertException
  Expected: ' Infinity ' to raise a DecimalError.
  Actual:   no exception, result: Infinity
  Test number 875 failed: assertException
  Expected: 'Infinity ' to raise a DecimalError.
  Actual:   no exception, result: Infinity
  Test number 876 failed: assertException
  Expected: ' Infinity ' to raise a DecimalError.
  Actual:   no exception, result: Infinity
  Test number 877 failed: assertException
  Expected: ' -Infinity' to raise a DecimalError.
  Actual:   no exception, result: -Infinity
  Test number 878 failed: assertException
  Expected: ' +Infinity' to raise a DecimalError.
  Actual:   no exception, result: Infinity
  Test number 879 failed: assertException
  Expected: '.Infinity' to raise a DecimalError.
  Actual:   no exception, result: NaN
  Test number 891 failed: assertException
  Expected: ' 0' to raise a DecimalError.
  Actual:   no exception, result: 0
  Test number 892 failed: assertException
  Expected: '0 ' to raise a DecimalError.
  Actual:   no exception, result: 0
  Test number 893 failed: assertException
  Expected: ' 0 ' to raise a DecimalError.
  Actual:   no exception, result: 0
  Test number 894 failed: assertException
  Expected: '0-' to raise a DecimalError.
  Actual:   no exception, result: NaN
  Test number 895 failed: assertException
  Expected: ' -0' to raise a DecimalError.
  Actual:   no exception, result: 0
  Test number 896 failed: assertException
  Expected: '-0 ' to raise a DecimalError.
  Actual:   no exception, result: 0
  Test number 897 failed: assertException
  Expected: '+0 ' to raise a DecimalError.
  Actual:   no exception, result: 0
  Test number 898 failed: assertException
  Expected: ' +0' to raise a DecimalError.
  Actual:   no exception, result: 0
  Test number 899 failed: assertException
  Expected: ' .0' to raise a DecimalError.
  Actual:   no exception, result: 0
  Test number 900 failed: assertException
  Expected: '0. ' to raise a DecimalError.
  Actual:   no exception, result: 0
  Test number 901 failed: assertException
  Expected: '+-0' to raise a DecimalError.
  Actual:   no exception, result: NaN
  Test number 902 failed: assertException
  Expected: '-+0' to raise a DecimalError.
  Actual:   no exception, result: NaN
  Test number 903 failed: assertException
  Expected: '--0' to raise a DecimalError.
  Actual:   no exception, result: NaN
  Test number 904 failed: assertException
  Expected: '++0' to raise a DecimalError.
  Actual:   no exception, result: NaN
  Test number 905 failed: assertException
  Expected: '.-0' to raise a DecimalError.
  Actual:   no exception, result: NaN
  Test number 906 failed: assertException
  Expected: '.+0' to raise a DecimalError.
  Actual:   no exception, result: NaN
  Test number 907 failed: assertException
  Expected: '0 .' to raise a DecimalError.
  Actual:   no exception, result: NaN
  Test number 908 failed: assertException
  Expected: '. 0' to raise a DecimalError.
  Actual:   no exception, result: NaN
  Test number 909 failed: assertException
  Expected: '..0' to raise a DecimalError.
  Actual:   no exception, result: NaN
  Test number 910 failed: assertException
  Expected: '+.-0' to raise a DecimalError.
  Actual:   no exception, result: NaN
  Test number 911 failed: assertException
  Expected: '-.+0' to raise a DecimalError.
  Actual:   no exception, result: NaN
  Test number 912 failed: assertException
  Expected: '+. 0' to raise a DecimalError.
  Actual:   no exception, result: NaN
  Test number 913 failed: assertException
  Expected: '.0.' to raise a DecimalError.
  Actual:   no exception, result: NaN
  Test number 931 failed: assertException
  Expected: ' 1' to raise a DecimalError.
  Actual:   no exception, result: 1
  Test number 932 failed: assertException
  Expected: '1 ' to raise a DecimalError.
  Actual:   no exception, result: 1
  Test number 933 failed: assertException
  Expected: ' 1 ' to raise a DecimalError.
  Actual:   no exception, result: 1
  Test number 934 failed: assertException
  Expected: '1-' to raise a DecimalError.
  Actual:   no exception, result: NaN
  Test number 935 failed: assertException
  Expected: ' -1' to raise a DecimalError.
  Actual:   no exception, result: -1
  Test number 936 failed: assertException
  Expected: '-1 ' to raise a DecimalError.
  Actual:   no exception, result: -1
  Test number 937 failed: assertException
  Expected: ' +1' to raise a DecimalError.
  Actual:   no exception, result: 1
  Test number 938 failed: assertException
  Expected: '+1' to raise a DecimalError.
  Actual:   no exception, result: 1
  Test number 939 failed: assertException
  Expected: '.1.' to raise a DecimalError.
  Actual:   no exception, result: NaN
  Test number 940 failed: assertException
  Expected: '+-1' to raise a DecimalError.
  Actual:   no exception, result: NaN
  Test number 941 failed: assertException
  Expected: '-+1' to raise a DecimalError.
  Actual:   no exception, result: NaN
  Test number 942 failed: assertException
  Expected: '--1' to raise a DecimalError.
  Actual:   no exception, result: NaN
  Test number 943 failed: assertException
  Expected: '++1' to raise a DecimalError.
  Actual:   no exception, result: NaN
  Test number 944 failed: assertException
  Expected: '.-1' to raise a DecimalError.
  Actual:   no exception, result: NaN
  Test number 945 failed: assertException
  Expected: '.+1' to raise a DecimalError.
  Actual:   no exception, result: NaN
  Test number 946 failed: assertException
  Expected: '1 .' to raise a DecimalError.
  Actual:   no exception, result: NaN
  Test number 947 failed: assertException
  Expected: '. 1' to raise a DecimalError.
  Actual:   no exception, result: NaN
  Test number 948 failed: assertException
  Expected: '..1' to raise a DecimalError.
  Actual:   no exception, result: NaN
  Test number 949 failed: assertException
  Expected: '+.-1' to raise a DecimalError.
  Actual:   no exception, result: NaN
  Test number 950 failed: assertException
  Expected: '-.+1' to raise a DecimalError.
  Actual:   no exception, result: NaN
  Test number 951 failed: assertException
  Expected: '+. 1' to raise a DecimalError.
  Actual:   no exception, result: NaN
  Test number 952 failed: assertException
  Expected: '-. 1' to raise a DecimalError.
  Actual:   no exception, result: NaN
  Test number 953 failed: assertException
  Expected: '1..' to raise a DecimalError.
  Actual:   no exception, result: NaN
  Test number 954 failed: assertException
  Expected: '+1..' to raise a DecimalError.
  Actual:   no exception, result: NaN
  Test number 955 failed: assertException
  Expected: '-1..' to raise a DecimalError.
  Actual:   no exception, result: NaN
  Test number 956 failed: assertException
  Expected: '-.1.' to raise a DecimalError.
  Actual:   no exception, result: NaN
  Test number 957 failed: assertException
  Expected: '+.1.' to raise a DecimalError.
  Actual:   no exception, result: NaN
  Test number 958 failed: assertException
  Expected: '.-10.' to raise a DecimalError.
  Actual:   no exception, result: NaN
  Test number 959 failed: assertException
  Expected: '.+10.' to raise a DecimalError.
  Actual:   no exception, result: NaN
  Test number 960 failed: assertException
  Expected: '. 10.' to raise a DecimalError.
  Actual:   no exception, result: NaN
  Test number 967 failed: assertException
  Expected: 'undefined' to raise a DecimalError.
  Actual:   no exception, result: NaN
  Test number 969 failed: assertException
  Expected: 'null' to raise a DecimalError.
  Actual:   no exception, result: NaN
  Test number 975 failed: assertException
  Expected: '' to raise a DecimalError.
  Actual:   no exception, result: NaN
  Test number 976 failed: assertException
  Expected: ' ' to raise a DecimalError.
  Actual:   no exception, result: NaN
  Test number 978 failed: assertException
  Expected: '23e' to raise a DecimalError.
  Actual:   no exception, result: NaN
  Test number 979 failed: assertException
  Expected: 'e4' to raise a DecimalError.
  Actual:   no exception, result: NaN
  Test number 980 failed: assertException
  Expected: 'ff' to raise a DecimalError.
  Actual:   no exception, result: NaN
  Test number 981 failed: assertException
  Expected: 'oxg' to raise a DecimalError.
  Actual:   no exception, result: NaN
  Test number 982 failed: assertException
  Expected: '0Xfi' to raise a DecimalError.
  Actual:   no exception, result: NaN
  Test number 983 failed: assertException
  Expected: '++45' to raise a DecimalError.
  Actual:   no exception, result: NaN
  Test number 984 failed: assertException
  Expected: '--45' to raise a DecimalError.
  Actual:   no exception, result: NaN
  Test number 985 failed: assertException
  Expected: '9.99--' to raise a DecimalError.
  Actual:   no exception, result: NaN
  Test number 986 failed: assertException
  Expected: '9.99++' to raise a DecimalError.
  Actual:   no exception, result: NaN
  Test number 987 failed: assertException
  Expected: '0 0' to raise a DecimalError.
  Actual:   no exception, result: NaN
 907 of 987 tests passed in 7 ms
 Testing dividedBy... 1000 of 1000 tests passed in 11 ms
 Testing dividedToIntegerBy... 1000 of 1000 tests passed in 9 ms
 Testing decimalPlaces and precision... 133 of 133 tests passed in 1 ms
 Testing exp... 100 of 100 tests passed in 11 ms
 Testing floor... 100 of 100 tests passed in 0 ms
 Testing hypot... 100 of 100 tests passed in 6 ms
 Testing immutability... 3341 of 3341 tests passed in 91 ms
 Testing integer pow...
  Test number 493 failed: assertException
  Expected: .pow('') to raise a DecimalError.
  Actual:   no exception, result: NaN
  Test number 494 failed: assertException
  Expected: .pow(' ') to raise a DecimalError.
  Actual:   no exception, result: NaN
  Test number 495 failed: assertException
  Expected: .pow('4e') to raise a DecimalError.
  Actual:   no exception, result: NaN
  Test number 496 failed: assertException
  Expected: .pow('hello') to raise a DecimalError.
  Actual:   no exception, result: NaN
  Test number 497 failed: assertException
  Expected: .pow('      ') to raise a DecimalError.
  Actual:   no exception, result: NaN
 495 of 500 tests passed in 6 ms
 Testing isFinite, isInteger, isNaN, isNegative, isZero, isDecimal... 214 of 214 tests passed in 1 ms
 Testing ln... 499 of 499 tests passed in 74 ms
 Testing log... 200 of 200 tests passed in 40 ms
 Testing log10... 500 of 500 tests passed in 63 ms
 Testing log2... 200 of 200 tests passed in 58 ms
 Testing min and max... 100 of 100 tests passed in 0 ms
 Testing minus... 1000 of 1000 tests passed in 5 ms
 Testing modulo... 1000 of 1000 tests passed in 32 ms
 Testing negated... 200 of 200 tests passed in 0 ms
 Testing plus... 1000 of 1000 tests passed in 5 ms
 Testing pow... 108 of 108 tests passed in 27 ms
 Testing random... 1000 of 1000 tests passed in 4 ms
 Testing round... 500 of 500 tests passed in 2 ms
 Testing sign... 16 of 16 tests passed in 0 ms
 Testing sin... 100 of 100 tests passed in 46 ms
 Testing sinh... 100 of 100 tests passed in 19 ms
 Testing sqrt... 1000 of 1000 tests passed in 26 ms
 Testing sum... 37 of 37 tests passed in 0 ms
 Testing tan... 100 of 100 tests passed in 47 ms
 Testing tanh... 100 of 100 tests passed in 36 ms
 Testing times... 1000 of 1000 tests passed in 5 ms
 Testing toBinary... 327 of 327 tests passed in 11 ms
 Testing toDecimalPlaces... 500 of 500 tests passed in 2 ms
 Testing toExponential... 500 of 500 tests passed in 2 ms
 Testing toFixed... 500 of 500 tests passed in 1 ms
 Testing toFraction... 200 of 200 tests passed in 20 ms
 Testing toHex... 309 of 309 tests passed in 3 ms
 Testing toNearest... 157 of 157 tests passed in 0 ms
 Testing toNumber... 40 of 40 tests passed in 0 ms
 Testing toOctal... 293 of 293 tests passed in 3 ms
 Testing toPrecision... 500 of 500 tests passed in 3 ms
 Testing toSignificantDigits... 500 of 500 tests passed in 2 ms
 Testing toString... 338 of 338 tests passed in 1 ms
 Testing trunc... 100 of 100 tests passed in 1 ms
 Testing valueOf... 40 of 40 tests passed in 0 ms

 In total, 22555 of 22640 tests passed in 1.066 secs.
```

File: src/libs/safedecimal/index.ts [hunk 3]
```diff
@@ -0,0 +1,24 @@
+import { Decimal } from 'decimal.js';
+
+function sanitizeDecimal(value: Decimal.Value) {
```
[the remark this answers is not in the document]
Comment (tiagofilipenunes) [reply src/libs/safedecimal/index.ts:3]: Added

File: src/components/trade/tradeWidget/TradeWidgetBuySell.tsx [hunk 4]
```diff
@@ -70,7 +70,7 @@ export const TradeWidgetBuySell = (props: TradeWidgetBuySellProps) => {
         buy,
         buyToken: target,
         sellToken: source,
-        valueUsd: getFiatValueSource(sourceInput, true).toString(),
+        valueUsd: getFiatValueSource(targetInput, true).toString(),
```
Comment (zavelevsky) [inline src/components/trade/tradeWidget/TradeWidgetBuySell.tsx:73]: seems like a mistake - should be sourceInput?

File: src/components/trade/tradeWidget/useBuySell.ts [hunk 5]
```diff
@@ -112,33 +112,25 @@ export const useBuySell = ({
...
+    const checkTarget = () =>
+      new SafeDecimal(targetInput).gt(liquidityQuery.data || 0);
 
     const set = () => setIsLiquidityError(true);
     setIsLiquidityError(false);
 
     if (isTradeBySource) {
       if (checkSource()) {
         setTargetInput('...');
+        setRate('');
```
Comment (zavelevsky) [inline src/components/trade/tradeWidget/useBuySell.ts:127]: why are we mixing additional things in the PR?

Comment (tiagofilipenunes) [reply src/components/trade/tradeWidget/useBuySell.ts:127]: Removed

**Review by zavelevsky — APPROVED** [review]
LGTM

**Pull request merged**